### PR TITLE
[WIP] added friends

### PR DESCRIPTION
**Pull request by UnkDevE**
added the an equivalent to the friend design pattern in c++ to be used with/out meta. just a utility that will be used in lex-game. Creating wrappers - integrating in meta would be the next step.

### CODE REVIEW

**Comment by UnkDevE** [comment]
started writing it in lex-game-Engine

**Comment by UnkDevE** [comment]
will fix

**Comment by jamierocks** [comment]
1. `gradle licenseFormat`
2. Usage example

**Comment by UnkDevE** [comment]
done it manually

File: src/main/java/uk/jamierocks/meta/api/MetaQuery.java [hunk 1]
```diff
@@ -36,9 +36,7 @@ protected MetaQuery(String name) {
         this.name = name;
     }
 
-    public static MetaQuery of(String name) {
-        return new MetaQuery(name);
-    }
+    public static MetaQuery of(String name) {return new MetaQuery(name);}
```
Comment (jamierocks) [inline src/main/java/uk/jamierocks/meta/api/MetaQuery.java:39]: Please undo this.

File: src/main/java/uk/jamierocks/meta/impl/friend/Friend.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        return FriendsOf.contains(F.getIdentity().getType());
+    }
+
+    Identity getIdentity(){
+        return  id;
+    }
+
+    static private Identity id;
+
+    static private Set<String> FriendsOf;
```
Comment (jamierocks) [inline src/main/java/uk/jamierocks/meta/impl/friend/Friend.java]: This should be called friendsOf.

**Comment by UnkDevE** [comment]
ok thought so.

File: src/main/java/uk/jamierocks/meta/impl/friend/Friend.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    public boolean isFriend(Friend F){
+        return FriendsOf.contains(F.getIdentity().getType());
+    }
+
+    Identity getIdentity(){
+        return  id;
+    }
+
+    static private Identity id;
+
```
Comment (jamierocks) [inline src/main/java/uk/jamierocks/meta/impl/friend/Friend.java:34]: I can't see any reason why these need to be static.

Also if it does it should be `private static ...`

File: src/main/java/uk/jamierocks/meta/impl/friend/Identity.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
```
Comment (jamierocks) [inline src/main/java/uk/jamierocks/meta/impl/friend/Identity.java]: javadocs

File: src/main/java/uk/jamierocks/meta/impl/friend/Friend.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+package uk.jamierocks.meta.impl.friend;
+
+import java.lang.reflect.Type;
+import java.util.Set;
+
```
Comment (jamierocks) [inline src/main/java/uk/jamierocks/meta/impl/friend/Friend.java:24]: javadocs

**Comment by jamierocks** [comment]
I think the most important thing is an usage example.

**Comment by UnkDevE** [comment]
added that too.

**Comment by UnkDevE** [comment]
doing docs now

File: src/main/java/uk/jamierocks/meta/impl/friend/Identity.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
+package uk.jamierocks.meta.impl.friend;
+
+/*
```
Comment (jamierocks) [inline src/main/java/uk/jamierocks/meta/impl/friend/Identity.java]: License is in the wrong place :/

**Comment by UnkDevE** [comment]
wait forgot to do gradle bearwith

File: src/main/java/uk/jamierocks/meta/impl/friend/Friend.java (hunk 3 above)
Comment (UnkDevE) [inline src/main/java/uk/jamierocks/meta/impl/friend/Friend.java:34]: it should be _per_ class an identity, although your right it does not need to be, it just makes more sense(for me)

File: src/main/java/uk/jamierocks/meta/impl/friend/Friend.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+public abstract class Friend {
+
+    public Friend(){
+        id = new Identity(this);
+    }
+
+    /**
+     * adds a friend to the class
```
Comment (jamierocks) [inline src/main/java/uk/jamierocks/meta/impl/friend/Friend.java]: Upper case characters and full stops m8 xD

File: src/main/java/uk/jamierocks/meta/impl/friend/Friend.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+     * @param F friend to add
+     */
+    public void addFriend(Friend F){
+        FriendsOf.add(F.getIdentity().getType());
+    }
+
+    /**
+     * @param F Friend to check
+     * @returns true if F is a friend to this.
```
Comment (jamierocks) [inline src/main/java/uk/jamierocks/meta/impl/friend/Friend.java]: {@code true}

File: src/main/java/uk/jamierocks/meta/impl/friend/FriendValue.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package uk.jamierocks.meta.impl.friend;
+
+import uk.jamierocks.meta.api.value.Value;
+
+public abstract class FriendValue<V> extends Friend implements Value<V> {
+
```
Comment (jamierocks) [inline src/main/java/uk/jamierocks/meta/impl/friend/FriendValue.java]: /**
  * {@inheritDocs}
  */

File: src/main/java/uk/jamierocks/meta/impl/friend/Friend.java (hunk 7 above)
Comment (UnkDevE) [inline src/main/java/uk/jamierocks/meta/impl/friend/Friend.java]: will fix

Comment (UnkDevE) [inline src/main/java/uk/jamierocks/meta/impl/friend/Friend.java]: fixed

File: src/main/java/uk/jamierocks/meta/impl/friend/Friend.java (hunk 8 above)
Comment (UnkDevE) [inline src/main/java/uk/jamierocks/meta/impl/friend/Friend.java]: ? use instead of @return ?

Comment (UnkDevE) [inline src/main/java/uk/jamierocks/meta/impl/friend/Friend.java]: maybe not. XD just got the joke.

File: src/main/java/uk/jamierocks/meta/impl/friend/FriendValue.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+        if(isFriend(F)) {
+            this.get();
+        }
+        else{
+            throw new IllegalArgumentException("not a friend of "+this.getClass().getName());
+        }
+        return null;
+    }
+
+}
```
Comment (jamierocks) [inline src/main/java/uk/jamierocks/meta/impl/friend/FriendValue.java]: From what I can tell having a FriendValue is pointless.
If it works like I think it does it should be a detail handled by the valueprocessors.

Comment (UnkDevE) [inline src/main/java/uk/jamierocks/meta/impl/friend/FriendValue.java]: that makes more sense like that; thanks.

Comment (UnkDevE) [inline src/main/java/uk/jamierocks/meta/impl/friend/FriendValue.java]: will do now

**Comment by jamierocks** [comment]
This PR is now closed.

It was decided that this functionality is not necessary to be part of Meta API.

File: src/main/java/uk/jamierocks/meta/impl/friend/Friend.java (hunk 8 above)
Comment (jamierocks) [inline src/main/java/uk/jamierocks/meta/impl/friend/Friend.java]: // I just noticed this - I am responding although the PR is closed

This wasn't a joke.
I was talking about replacing
`true if F is a friend to this.`
with
`{@code true} if F is a friend to this.`

**Pull request closed without merging**